var/spack/repos/builtin/packages/hipsycl/package.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+from os import path
+from llnl.util import filesystem
+
+
+class Hipsycl(CMakePackage):
+    """hipSYCL is an implementation of the SYCL standard programming model
+    over NVIDIA CUDA/AMD HIP"""
+
+    homepage = "https://github.com/illuhad/hipSYCL"
+    url = "https://github.com/illuhad/hipSYCL/archive/v0.8.0.tar.gz"
+    git = "https://github.com/illuhad/hipSYCL.git"
+
+    maintainers = ["nazavode"]
+
+    provides("sycl")
+
+    version("master", branch="master", submodules=True)
+    version(
+        "0.8.0",
+        commit="2daf8407e49dd32ebd1c266e8e944e390d28b22a",
+        submodules=True,
+    )
+
+    variant(
+        "cuda",
+        default=False,
+        description="Enable CUDA backend for SYCL kernels",
+    )
+
+    depends_on("cmake@3.5:", type="build")
+    depends_on("boost +filesystem")
+    depends_on("python@3:")
+    depends_on("llvm@8: +clang", when="~cuda")
+    depends_on("llvm@9: +clang", when="+cuda")
+    # LLVM PTX backend requires cuda7:10.1.9999 (https://tinyurl.com/v82k5qq)
+    depends_on("cuda@9:10.1.9999", when="@0.8.1: +cuda")
+    # hipSYCL@:0.8.0 requires cuda@9:10.0.9999 due to a known bug
+    depends_on("cuda@9:10.0.9999", when="@:0.8.0 +cuda")
+
+    conflicts(
+        "^llvm build_type=Debug",
+        when="+cuda",
+        msg="LLVM debug builds don't work with hipSYCL CUDA backend; for "
+        "further info please refer to: "
+        "https://github.com/illuhad/hipSYCL/blob/master/doc/install-cuda.md",
+    )
+
+    def cmake_args(self):
+        spec = self.spec
+        args = [
+            "-DWITH_CPU_BACKEND:Bool=TRUE",
```

Comment:
The CPU backend uses the compiler that hipSYCL is compiled with as default compiler (users can still change it via command line). In spack, this compiler is e.g. `spack/env/gcc/g++` which can only be run from spack, so compilation for CPU with `syclcc` will fail unless the user specifies a different compiler using `--hipsycl-cpu-cxx`/`HIPSYCL_CPU_CXX`. To avoid this and make the default compiler work, I would suggest to also use the clang++ here that we already have available. There is no dedicated cmake option yet for this, it just uses the regular `CMAKE_CXX_COMPILER` as compiler for the CPU backend, so:
* we could just set the Cmake C++ compiler to clang++, which however might be unconventional for spack users since setting the spack compiler to be used for this package will not work as expected
* we could edit the generated `etc/hipSYCL/syclcc.json` file after compilation to point to our clang++

Also, the compiler for the CPU backend has to support OpenMP (and C++14). Do we need to add a dependency to make sure llvm libomp (or libgomp for gcc) is available?

Reply:
I came up with a patch that tries to address this issue by modifying the generated `syclcc.json` but I noticed that the resulting `syclcc` uses the Spack-specified build compiler (e.g.: `g++`) for `--hipsycl-platform=cpu` but falls back to `default-clang` (e.g. the `clang++` provided by the concretized `^llvm`) when `--hipsycl-platform=cuda`. Since the latter is unavoidable, I'm seriously thinking about making `hipsycl` in Spack buildable only by `%clang`. That would:
1. avoid *uncomfortable* situations where binaries compiled by `syclcc` link both `libstdc++` and `libc++`;
2. avoid *surprising* situations where `syclcc` emits objects produced by different compilers depending on the target platform;
3. allow us to remove `^llvm` as a dependency and just use the `%clang` build compiler.

What do you folks think? @illuhad @alalazo

Reply:
My stance is that this is a feature that hipSYCL is designed for -- users compiling for CPU might want to use some vendor compiler like Intel C++ or IBM xl compilers, and it makes sense for hipSYCL to allow for that, especially in an HPC context. On the other hand, when compiling for CUDA or ROCm, we have no choice but to use clang since we need a CUDA/HIP compiler and the clang plugin.
Personally, I don't see a problem as long as the behavior is documented, but of course I am used to this behavior. If this is undesired, we could:
* Always compile with clang and clearly tell the user that the CPU compiler can optionally be changed using `syclcc` `--hipsycl-cpu-cxx=` command line argument or `HIPSYCL_CPU_CXX` environment variable
* Always compile with clang, but allow the user to specify the compiler that goes into `syclcc.json` when installing (and default to clang). Out of the two alternative options, this would be my preferred one.

Reply:
@illuhad that makes total sense, I was missing the point that allowing a custom compiler for `--hipsycl-platform=cpu` is actually a design goal. I'm going to follow the path you were suggesting before of treating the Spack build compiler just as a default to be set in the installed configuration file. Thanks!

Reply:
Having thought about it some more, I think it would also make sense to
* allow only clang to be used as a compiler when enabling GPU backends and
* allow any compiler to be used when only compiling the CPU backend.

In this way, there are no surprises like `libstdc++/libc++` mismatches for users who have installed all backends, and, thanks to spack, users who want an optimized CPU backend using their favorite compiler can just `spack install` it.

Reply:
@illuhad I think I've been overly conservative regarding my proposal of restricting allowed compilers. With all the `rpath` issues settled, binaries compiled by `syclcc` seems to work quite smoothly even with some evil compiler combinations. Regarding linking both `libstdc++` and `libc++`: while we control the actual usage of the latter, the former is linked to all the binaries shipped by NVIDIA, so a soon as you're linking `cudart.so` for instance, you get `libstdc++`. I think we can just settle on the current state (allow every combination of compilers) and see how it works out for users.

Reply:
Regarding specifying the dependence on C++14: I don't think there is a way to actually enforce those unless you take compiler version as a proxy for language std support, while for OpenMP I actually have no idea since both `%gcc` and `%clang` provided by Spack build their own OpenMP runtimes by default without exposing any knob...pinging @alalazo here

Reply:
I see. The reason I brought that up is because there are still systems around that ship ancient compilers (e.g. CentOS 7/RHEL 7) and it would be nice if spack could detect the problem. However, I think that the build will just exist with a clear error message in that case since syclcc will automatically append `-std=c++14` to the command line. So, I don't think that's terribly important if it cannot be done easily in spack. Same for OpenMP.

Reply:
Added an additional check on the `%gcc` version being used in 66be3460826de24f42c1848ba44b9861737ad1db. Right now it is sort of a comment rather than a proper `conflict` since the `llvm` needed by `hipsycl` adds a stricter constraint during concretization:

```console
$ spack spec hipsycl %gcc@4.8.5
1. "%gcc@:4.9999" conflicts with "hipsycl" [hipSYCL needs proper C++14 support to be built, %gcc is too old]
2. "%gcc@:5.0.999" conflicts with "llvm@8:"
```

Since I'm one of those *lucky* users working on ancient platforms, I felt obliged to add it :sunglasses: